### PR TITLE
 Add support for cursor based pagination and remove page based pagination

### PR DESCRIPTION
**Pull request by daesu**
## What
Major version update to support cursor based pagination

## Description
- Added support for new cursor based pagination
- Removed deprecated structs
- Added integration tests for pagination endpoints

### CODE REVIEW

**Review by pkopac — DISMISSED** [review]
LGTM, but you need to provide migration howto/changelog. There's a bunch of renaming and an example of "iterate all pages" should be provided.

**Review by swember — DISMISSED** [review]

File: subscription_events.go [hunk 1]
```diff
@@ -53,7 +53,7 @@ type SubscriptionEventParams struct {
 	Params *SubscriptionEvent `json:"subscription_event"`
 }
 
-func (api API) ListSubscriptionEvents(filters *FilterSubscriptionEvents, cursor *MetaCursor) (*SubscriptionEvents, error) {
+func (api API) ListSubscriptionEvents(filters *FilterSubscriptionEvents, cursor *Cursor) (*SubscriptionEvents, error) {
```
Comment (kmossco) [inline subscription_events.go:56]: This won't let us run a cursored query that also has pagination filters like `per_page: 1` right? 

```go
type FilterSubscriptionEvents struct {
	CustomerExternalID     string `json:"customer_external_id,omitempty"`
	DataSourceUUID         string `json:"data_source_uuid,omitempty"`
	EffectiveDate          string `json:"effective_date,omitempty"`
	EventDate              string `json:"event_date,omitempty"`
	EventType              string `json:"event_type,omitempty"`
	ExternalID             string `json:"external_id,omitempty"`
	PlanExternalID         string `json:"plan_external_id,omitempty"`
	SubscriptionExternalID string `json:"subscription_external_id,omitempty"`
}
```

Comment (kmossco) [inline subscription_events.go:56]: There's something weird with this endpoint 🤔 If I do:

```go
spew.Dump(api.ListSubscriptionEvents(&cm.FilterSubscriptionEvents{
    CustomerExternalID: "cus_OL8zPfJI7phy9k",
}, &cm.Cursor{ PerPage: 1, }))
```

I get this result:

```go
(*chartmogul.SubscriptionEvents)(0x14000072ea0)({
 SubscriptionEvents: ([]*chartmogul.SubscriptionEvent) (len=1 cap=4) {
  (*chartmogul.SubscriptionEvent)(0x14000140300)({
   ID: (uint64) 374186133,
   DataSourceUUID: (string) (len=39) "ds_d02c7c42-2d1d-11ee-8e58-3fa5919351b4",
   CustomerExternalID: (string) (len=18) "cus_OL8zPfJI7phy9k",
   SubscriptionSetExternalID: (string) "",
   SubscriptionExternalID: (string) (len=17) "si_OL95DjqoHYm95u",
   PlanExternalID: (string) (len=30) "price_1NYSWgIFhx3XTAwSSnub7K43",
   EventDate: (string) (len=20) "2023-10-25T12:37:05Z",
   EffectiveDate: (string) (len=20) "2023-10-25T12:37:05Z",
   EventType: (string) (len=35) "subscription_cancellation_scheduled",
   ExternalID: (string) (len=17) "si_OL95DjqoHYm95u",
   Errors: (map[string]interface {}) {
   },
   CreatedAt: (string) (len=20) "2023-10-25T12:38:06Z",
   UpdatedAt: (string) (len=20) "2023-10-25T12:38:06Z",
   Quantity: (int32) 0,
   Currency: (string) "",
   AmountInCents: (int32) 0,
   TaxAmountInCents: (int32) 0,
   RetractedEventId: (string) ""
  })
 },
 Meta: (chartmogul.MetaCursor) {
  Pagination: (chartmogul.Pagination) {
   Cursor: (string) "",
   HasMore: (bool) false
  }
 }
})
(interface {}) <nil>
```

Note that the cursor is empty and `HasMore` returns false, but if I run this in cURL I get the right response:

```json
{
	"subscription_events": [
		{
			"id": 374186133,
			"data_source_uuid": "ds_d02c7c42-2d1d-11ee-8e58-3fa5919351b4",
			"customer_external_id": "cus_OL8zPfJI7phy9k",
			"subscription_set_external_id": null,
			"subscription_external_id": "si_OL95DjqoHYm95u",
			"plan_external_id": "price_1NYSWgIFhx3XTAwSSnub7K43",
			"event_date": "2023-10-25T12:37:05Z",
			"effective_date": "2023-10-25T12:37:05Z",
			"event_type": "subscription_cancellation_scheduled",
			"external_id": "si_OL95DjqoHYm95u",
			"errors": {},
			"created_at": "2023-10-25T12:38:06Z",
			"updated_at": "2023-10-25T12:38:06Z",
			"quantity": null,
			"currency": null,
			"amount_in_cents": null,
			"tax_amount_in_cents": null,
			"retracted_event_id": null
		}
	],
	"meta": {
		"next_key": 374186133,
		"prev_key": null,
		"before_key": "2023-11-09T15:09:52.899Z",
		"page": 1,
		"total_pages": 3
	},
	"cursor": "MjAyMy0xMC0yNVQxMjozODowNi44Njc0MTIwMDBaJjM3NDE4NjEzMw==",
	"has_more": true
}
```

**Comment by daesu** [comment]
Created new branch `v4` from PR.

**Pull request closed without merging**